ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+                tid <- myThreadId
+                result <- atomically $ do
+                        v <- readTVar valencyVar
+                        if v > 0 then do
+                                    addConnection tbl remoteAddr localAddr' [valencyVar]
+                                    writeTVar hasRefVar True
+                                    modifyTVar' conThreads (Set.delete tid)
+                                    if v == 1 then return ConnectSuccessLast
+                                              else return ConnectSuccess
+                                else return ConnectFail
+                traceWith tracer $ SubscriptionTraceConnectEnd remoteAddr result
+                case result of
+                    ConnectSuccess -> k sd
+                    ConnectSuccessLast -> do
+                            outstandingConThreads <- atomically $ readTVar conThreads
+                            mapM_ (\a -> throwTo a
+                                   (SubscriberError SubscriberParrallelConnectionCancelled
+                                    "Parrallel connection cancelled"
+                                    callStack)) outstandingConThreads
+                            k sd
+                    ConnectFail -> return ()
+
+            )
+
+data IPSubscriptionTarget = IPSubscriptionTarget {
+      ispIps     :: ![Socket.SockAddr]
+    , ispValency :: !Int
+    } deriving (Eq, Show)
+
+ipSubscriptionWorker
+    :: ConnectionTable
+    -> Tracer IO (WithIPList SubscriptionTrace)
+    -> Maybe Socket.SockAddr
+    -> Maybe Socket.SockAddr
+    -> (Socket.SockAddr -> Maybe DiffTime)
+    -> IPSubscriptionTarget
+    -> (Socket.Socket -> IO ())
+    -> (Async () -> IO t)
+    -> IO t
+ipSubscriptionWorker tbl tracer localIPv4 localIPv6 connectionAttemptDelay ips cb k =
+    subscriptionWorker
+            tbl
+            (ipListTracer localIPv4 localIPv6 (ispIps ips) tracer)
+            localIPv4
+            localIPv6
+            connectionAttemptDelay
+            (return $ listSubscriptionTarget (ispIps ips))
+            (ispValency ips)
+            cb k
+
+subscriptionWorker
+    :: ConnectionTable
+    -> Tracer IO SubscriptionTrace
+    -> Maybe Socket.SockAddr
+    -> Maybe Socket.SockAddr
+    -> (Socket.SockAddr -> Maybe DiffTime)
+    -> IO (SubscriptionTarget IO Socket.SockAddr)
+    -> Int
+    -> (Socket.Socket -> IO ())
+    -> (Async () -> IO t)
+    -> IO t
+subscriptionWorker tbl tracer localIPv4 localIPv6 connectionAttemptDelay getTargets valency
+        cb k = do
+    childrenVar <- newTVarM Set.empty
+    withAsync (worker childrenVar)
+              (\aid -> (k aid) `finally` (do
+                  children <- atomically $ readTVar childrenVar
+                  mapM_ (\a -> throwTo a
+                                   (SubscriberError SubscriberWorkerCancelled
+                                    "SubscriptionWorker exiting"
+                                    callStack)) $ Set.toList children
```

Comment:
Yeah, we should use a separately implemented concurrency abstraction for this thread registry.